torch/fx/experimental/proxy_tensor.py

```diff
@@ -284,7 +286,17 @@ def __repr__(self):
 
     @classmethod
     def __torch_dispatch__(cls, func_overload, types, args=(), kwargs=None):
-        return proxy_call(func_overload, args, kwargs)
+        # Verify that the proxy mode for all arguments is the same
+        proxy_mode = None
+        for arg in pytree.tree_flatten((args, kwargs))[0]:
+            if isinstance(arg, ProxyTensor):
+                if proxy_mode is None:
+                    proxy_mode = arg.proxy_mode
+                else:
+                    assert proxy_mode is arg.proxy_mode, "All arguments must be in the same proxy mode"
+
+        with enable_torch_dispatch_mode(proxy_mode):
```

Comment:
Looks like this is latent bug from the fake tensor code; does `proxy_mode.restore()` work?

Reply:
When using `proxy_mode.restore()` `TestProxyTensorCPU.test_decomposition_interpreter_cpu` fails with
```py
RuntimeError: <torch.fx.experimental.proxy_tensor.ProxyTorchDispatchMode object at 0x7f084183ddb0> does not have any ancestors. Use the standard version instead of restore
```

Reply:
this fixes this error

```
diff --git a/torch/fx/experimental/proxy_tensor.py b/torch/fx/experimental/proxy_tensor.py
index a7e38074eb6..d5bafefc0f2 100644
--- a/torch/fx/experimental/proxy_tensor.py
+++ b/torch/fx/experimental/proxy_tensor.py
@@ -476,6 +476,9 @@ class DecompositionInterpreter(torch.fx.Interpreter):
         return out
 
     def run(self, *args, **kwargs):
+        # snapshot the current mode stack
+        with self.mode:
+            pass
         with decompose(self.decomposition_table):
             return super().run(*args, **kwargs)
 
```

I need to think more about what went wrong on the API design side though.

cc @samdow: what's going on here is people are allocating mode objects, but then never actually entering them (and manually calling API functions that make use of the mode object.)